eos-paygd/main.c

```diff
@@ -24,18 +24,79 @@
 #include <libgsystemservice/service.h>
 #include <libeos-payg/service.h>
 #include <signal.h>
+#include <linux/reboot.h>
+#include <sys/reboot.h>
 
+/* Force a reboot in situations where we are not able to enforce PAYG */
+static gboolean
+sync_and_reboot (gpointer user_data)
+{
+    g_warning ("bcsn: Forcing poweroff now!");
+    sync ();
+    reboot (LINUX_REBOOT_CMD_POWER_OFF);
+    return FALSE;
+}
 
 int
 main (int   argc,
       char *argv[])
 {
   g_autoptr(GError) error = NULL;
   g_autoptr(EpgService) service = NULL;
+  guint timeout_id;
+
+  /* If eos-paygd is running from the initramfs, change the process name so
+   * that it survives the pivot to the final root filesystem. This is an
+   * (ab)use of the functionality systemd has to support storage daemons needed
+   * for mounting/unmounting the root filesystem. See
+   * https://phabricator.endlessm.com/T27037
+   */
+  if (access ("/etc/initrd-release", F_OK) >= 0)
```

Comment:
Could this be made testable by adding a `--prefix` argument to `eos-paygd` which gets prepended to paths like these? So you could test it by running (or, having the tests run) `eos-paygd --prefix /tmp/test-initrd-root` and creating `etc/initrd-release` (etc.) within that? The other side of testing it would be overriding the `reboot()` syscall by having the test `LD_PRELOAD` a library which defines `reboot()` into the `eos-paygd` it runs.

(Not saying that all of this should be done in this PR, but I think it would make sense to have a plan, rather than just dropping the tests without a plan.)

Reply:
I opened https://phabricator.endlessm.com/T27360 on this topic